### PR TITLE
Collapsing table cohortpeople

### PR DESCRIPTION
**Pull request by EDsCODE**
## Changes

*Please describe.*  
- use a collapsingmergetree table to precalculate and store cohort <-> person mappings so that they can be utilized in queries for better filtering performance
- follow-up: once we merge this, get the table populated, performance test, then we proceed with updating the queries to use this table. There are cases to handle such as triggering calculations or live running the query if the table hasn't been populated with the cohort mappings
*If this affects the frontend, include screenshots.*  

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious

### CODE REVIEW

File: ee/clickhouse/migrations/0010_cohortpeople.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+from infi.clickhouse_orm import migrations
+
+from ee.clickhouse.sql.cohort import CREATE_COHORTPEOPLE_TABLE_SQL
+
+operations = [
+    migrations.RunSQL(CREATE_COHORTPEOPLE_TABLE_SQL),
```
Comment (macobo) [inline ee/clickhouse/migrations/0010_cohortpeople.py:6]: Let's pull this out to a separate PR - just for deployment saneness

File: ee/clickhouse/models/cohort.py [hunk 2]
```diff
@@ -17,14 +17,14 @@
...
-def format_person_query(cohort: Cohort) -> Tuple[str, Dict[str, Any]]:
+def format_person_query(cohort: Cohort, custom_match_field="person_id") -> Tuple[str, Dict[str, Any]]:
     filters = []
     params: Dict[str, Any] = {}
 
     if cohort.is_static:
         return (
-            "person_id IN (SELECT person_id FROM {} WHERE cohort_id = %(cohort_id)s AND team_id = %(team_id)s)".format(
-                PERSON_STATIC_COHORT_TABLE
+            "{match_field} IN (SELECT person_id FROM {static_cohort} WHERE cohort_id = %(cohort_id)s AND team_id = %(team_id)s)".format(
```
Comment (macobo) [inline ee/clickhouse/models/cohort.py:26]: Q: I think this would be nicer with an f-string

File: ee/clickhouse/models/cohort.py [hunk 3]
```diff
@@ -17,14 +17,14 @@
 from posthog.models import Action, Cohort, Filter, Team
 
 
-def format_person_query(cohort: Cohort) -> Tuple[str, Dict[str, Any]]:
+def format_person_query(cohort: Cohort, custom_match_field="person_id") -> Tuple[str, Dict[str, Any]]:
```
Comment (macobo) [inline ee/clickhouse/models/cohort.py:20]: Do we want to allow any string as a custom match field? If not, let's create a union type for it above.

File: posthog/models/cohort.py [hunk 4]
```diff
@@ -115,6 +115,12 @@ def calculate_people(self, use_clickhouse=is_clickhouse_enabled()):
             self.save()
             capture_exception(err)
 
+    def calculate_people_ch(self):
```
Comment (macobo) [inline posthog/models/cohort.py:118]: Q: Why is this not under `.calculate_people`? It introduces complexity below in various celery tasks, which feels weird.

File: ee/clickhouse/sql/person.py [hunk 5]
```diff
@@ -220,7 +220,7 @@
 """
 
 DELETE_PERSON_BY_ID = """
-INSERT INTO person (id, is_deleted) SELECT %(id)s, 1
+INSERT INTO person (id, team_id, is_deleted) SELECT %(id)s, %(team_id)s, 1
```
Comment (macobo) [inline ee/clickhouse/sql/person.py:223]: Q: Is this consistent with how plugin server does it - if not, please sync it up!

File: ee/clickhouse/models/cohort.py [hunk 6]
```diff
@@ -107,3 +107,45 @@ def insert_static_cohort(person_uuids: List[Optional[uuid.UUID]], cohort_id: int
...
+        AND {cohort_filter}
+    """.format(
+        cohort_filter=cohort_filter
+    )
+
+    sync_execute(
+        INSERT_PEOPLE_MATCHING_COHORT_ID_SQL, {**cohort_params, "cohort_id": cohort.pk, "team_id": cohort.team_id}
+    )
+
+    REMOVE_PEOPLE_NOT_MATCHING_COHORT_ID_SQL = """
```
Comment (macobo) [inline ee/clickhouse/models/cohort.py:133]: In the past we've pulled large queries out to `clickhouse/sql` folder, do we not anymore?

File: ee/clickhouse/models/cohort.py [hunk 7]
```diff
@@ -107,3 +107,45 @@ def insert_static_cohort(person_uuids: List[Optional[uuid.UUID]], cohort_id: int
         for person_uuid in person_uuids
     )
     sync_execute(INSERT_PERSON_STATIC_COHORT, persons)
+
+
+def recalculate_cohortpeople(cohort: Cohort):
+    cohort_filter, cohort_params = format_person_query(cohort, "person.id")
```
Comment (macobo) [inline ee/clickhouse/models/cohort.py:113]: Nit: perhaps worth making `person.id` a kwarg `format_person_query(cohort, custom_match_field="person.id")`

File: ee/clickhouse/models/cohort.py (hunk 3 above)
Comment (EDsCODE) [reply ee/clickhouse/models/cohort.py]: I just changed the param to a kwarg. Should it still be a union type?

File: posthog/models/cohort.py (hunk 4 above)
Comment (EDsCODE) [reply posthog/models/cohort.py:118]: I thought of doing this but splitting it out would allow the tasks to be distributed across different workers I think so that we wouldn't pile up a few expensive queries on the same worker.

Comment (macobo) [reply posthog/models/cohort.py:118]: Can you elaborate on this point - what's the plan with the other method, why do we still need it at all? Will we be getting rid of it at some point?

Comment (EDsCODE) [reply posthog/models/cohort.py:118]: Oh, I think the confusion is that I opted to manage the split of postgres vs clickhouse in `posthog/tasks/calculate_cohort.py` line 35. In that function, I'm calling the `calculate_cohort` function and `calculate_cohort_ch` so that we can split these two calculations into different tasks. 

Once we validate the cohorts are populated correctly and the surrounding querying is implemented I can make it only fire off the task respective to the deployment (clickhouse or postgres). Or alternatively but similarly, we can handle the clickhouse vs postgres difference in the function above instead of in the task. For now, I'm just keeping both because we're still testing the clickhouse table.

**Comment by macobo** [comment]
Let's merge and ship once the dependent PR is live
